onnxruntime/core/framework/execution_frame.h

```diff
@@ -138,6 +139,27 @@ class ExecutionFrame final : public IExecutionFrame {
   // If the retrival is sucessful, this function returns true and false otherwise.
   bool TryGetInferredShape(int index, TensorShape& shape) const override;
 
+  // Return the size of virtual memory allocated in runtime.
+  // The memory is usually used for activations in forward and backward passes.
+  const std::unordered_map<std::string, size_t> GetDynamicMemorySizeInfo() {
+    // Dynamic memory dictionary can be accessed by multiple threads when using
+    // parallel executor, so we need to lock the global mutex of ExecutionFrame
+    // before accessing it.
+    std::unique_lock<std::mutex> lock(mtx_);
```

Comment:
>std::unique_lock<std::mutex> lock(mtx_); [](start = 4, length = 40)

I think the mutex should be a member and shared by read and write

Reply:
I have a similar question for another comment. Does it mean we will implement ThreadSafeUnorderedMap? It's the best way to always put mutex and read/write together.

---
In reply to: [466616056](https://github.com/microsoft/onnxruntime/pull/4712#discussion_r466616056) [](ancestors = 466616056)

Reply:
The lock here does not protect anything. If the read is supposed to be not overlap in time with write, then maybe just remove the lock in read, while keeping the one in write?

---
In reply to: [466639446](https://github.com/microsoft/onnxruntime/pull/4712#discussion_r466639446) [](ancestors = 466639446,466616056)

Reply:
This `mtx_` is a member of this class.
```c++
  // Mutex which should be acquired when executing non-thread-safe member functions.
  // A current example is the tracker of dynamic memory allocation.
  mutable std::mutex mtx_;
```
Do we still need other change?

---
In reply to: [466674331](https://github.com/microsoft/onnxruntime/pull/4712#discussion_r466674331) [](ancestors = 466674331,466639446,466616056)

Reply:
You are right, let's keep this way

---
In reply to: [466683334](https://github.com/microsoft/onnxruntime/pull/4712#discussion_r466683334) [](ancestors = 466683334,466674331,466639446,466616056)